src/main.cpp

```diff
@@ -6355,7 +6355,26 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
 
         bool fCanDirectFetch = CanDirectFetch(chainparams.GetConsensus());
         CNodeState *nodestate = State(pfrom->GetId());
-        nodestate->fFirstHeadersReceived = true;
+
+        // During the initial peer handshake we must receive the initial headers which should be greater
+        // than or equal to our block height at the time of requesting GETHEADERS. This is because the peer has
+        // advertised a height >= to our own. Furthermore, because the headers max returned is as much as 2000 this
+        // could not be a mainnet re-org.
+        if (!nodestate->fFirstHeadersReceived)
+        {
+            // We want to make sure that the peer doesn't just send us any old valid header. The last header block
+            // hash they send us should now be equal to our current pindexLast block hash.
+            if (pindexLast && pindexLast->IsValid(BLOCK_VALID_TREE) &&
+                chainActive.Tip()->nChainWork <= pindexLast->nChainWork)
+            {
```

Comment:
Another reason to test for the claim the peer made - our request was based on our state at that time.  But here we're basing it on our state now.  We might have connected a block or 2 in the meantime, and this code is essentially finding fault with the peer for that.

Maintaining track of requests and whether replies are inconsistent or incomplete should be handled by the RM.